### PR TITLE
FEAT: Enable running Modin in remote context

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
Enables running Modin on automatically-spawned AWS cluster.

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves https://github.com/modin-project/modin/issues/1701
- [x] tests added and passing

### CODE REVIEW

**Comment by vnlitvinov** [comment]
Note that until this PR is merged and released to PyPI, we need to customize `ray_autoscaler.yml` to install Modin from source.

My current way of doing it is this:
```diff
diff --git a/modin/experimental/cloud/ray-autoscaler.yml b/modin/experimental/cloud/ray-autoscaler.yml
index da85bbc..dbf301f 100644
--- a/modin/experimental/cloud/ray-autoscaler.yml
+++ b/modin/experimental/cloud/ray-autoscaler.yml
@@ -123,7 +123,14 @@ setup_commands:
         conda activate modin
         conda install python==3.7.6

-        pip install modin "ray==0.8.6"
+        pip install "ray==0.8.6" s3fs
+        # need to install modin from source for now
+        rm -rf modin
+        git clone --single-branch --depth 1 --branch merge-hedgehog-and-snake https://github.com/vnlitvinov/modin.git
+        (cd modin && pip install -e .)
+
+
+
         echo 'export MODIN_RAY_CLUSTER=True' >> ~/.bashrc

     # Consider uncommenting these if you also want to run apt-get commands during setup

```

**Comment by vnlitvinov** [comment]
I've converted the PR to draft until I figure out how to fix broken `test_api`.

**Comment by vnlitvinov** [comment]
Note to reviewers: rebased on top of current `master` to get rid of CI failures.

**Comment by vnlitvinov** [comment]
Seems I have a few other tests that fail due to my changes, so converting back to draft again.

File: modin/pandas/__init__.py [hunk 1]
```diff
@@ -185,6 +163,53 @@ def _update_engine(publisher: Publisher):
...
+    pivot_table,
+    notnull,
+    notna,
+    pivot,
+    to_numeric,
+    unique,
+    value_counts,
+)
+from .plotting import Plotting as plotting
+
```
Comment (anmyachev) [inline modin/pandas/__init__.py:212]: Plz, make a separate PR for this.

**Comment by anmyachev** [comment]
Well, the main question is, can we get the same functionality, but without creating wrappers over entire Modin classes?

File: examples/cluster/mortgage-runner.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+
+# pip install git+https://github.com/intel-go/ibis.git@develop
+
+# NOTE: expects https://github.com/intel-go/omniscripts/tree/modin-rpyc-test checked out and in PYTHONPATH
```
Comment (anmyachev) [inline examples/cluster/mortgage-runner.py:17]: Is this in omniscripts master already?

**Comment by anmyachev** [comment]
I have found that `CannotSpawnCluster` doesn't print `self.cause` value. `ClusterError.__str__` method should be redefined I suppose.

File: examples/cluster/mortgage-runner.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+# the following import turns on experimental mode in Modin,
+# including enabling running things in remote cloud
+import modin.experimental.pandas as pd  # noqa: F401
+from modin.experimental.cloud import create_cluster, get_connection
+
+from mortgage import run_benchmark
+from mortgage.mortgage_pandas import etl_pandas
+
+cl = create_cluster("aws", "aws_credentials", cluster_name="rayscale-test")
```
Comment (devin-petersohn) [inline examples/cluster/mortgage-runner.py:27]:
```suggestion
test_cluster = create_cluster("aws", "aws_credentials", cluster_name="rayscale-test")
```

File: examples/cluster/mortgage-runner.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+with cl:
```
Comment (devin-petersohn) [inline examples/cluster/mortgage-runner.py:29]:
```suggestion
with test_cluster:
```

File: modin/__init__.py [hunk 5]
```diff
@@ -85,19 +85,19 @@ class Publisher(object):
     def __init__(self, name, value):
         self.name = name
         self.__value = value.title()
-        self.__subs = set()
-        self.__once = collections.defaultdict(set)
+        self.__subs = []
```
Comment (devin-petersohn) [inline modin/__init__.py:88]: Do we no longer have unique values? Or do we need just need order?

File: modin/__init__.py [hunk 6]
```diff
@@ -85,19 +85,19 @@ class Publisher(object):
...
+        self.__subs.append(callback)
         callback(self)
 
     def once(self, onvalue, callback):
         onvalue = onvalue.title()
         if onvalue == self.__value:
             callback(self)
         else:
-            self.__once[onvalue].add(callback)
+            self.__once[onvalue].append(callback)
```
[the remark this answers is not in the document]
Comment (devin-petersohn) [reply modin/__init__.py:100]: `append` is for the `list` converstion instead of `set` on line 89. I have the same question about `set` to `list` change.

File: modin/pandas/dataframe.py [hunk 7]
```diff
@@ -36,12 +36,15 @@
...
+from .groupby import DataFrameGroupBy
+
+from modin.experimental.cloud.meta_magic import make_wrapped_class
 
 
 @_inherit_docstrings(
     pandas.DataFrame, excluded=[pandas.DataFrame, pandas.DataFrame.__init__]
 )
-class DataFrame(BasePandasDataset):
+class _DataFrame(BasePandasDataset):
```
Comment (devin-petersohn) [inline modin/pandas/dataframe.py:47]: This is a pretty major change. Is monkey patching not possible?

File: modin/data_management/factories.py [hunk 8]
```diff
@@ -210,3 +210,43 @@ def prepare(cls):
         from modin.experimental.engines.pyarrow_on_ray.io import PyarrowOnRayIO
 
         cls.io_cls = PyarrowOnRayIO
+
+
+class ExperimentalPandasOnCloudrayFactory(ExperimentalBaseFactory):
```
Comment (anmyachev) [inline modin/data_management/factories.py:215]: Can you move this class into `modin/experimental` module?

File: modin/experimental/cloud/meta_magic.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import sys
+import inspect
+import types
+
+_SPECIAL = frozenset(("__new__", "__dict__"))
```
Comment (anmyachev) [inline modin/experimental/cloud/meta_magic.py:18]: Maybe `_LOCALS`?

File: modin/experimental/cloud/meta_magic.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    """
+
+    def __getattribute__(self, name):
+        if name in _SPECIAL:
+            # never proxy special attributes, always get them from the class type
+            res = object.__getattribute__(self, name)
+        else:
+            try:
+                # go for proxying class-level attributes first
+                res = object.__getattribute__(self, "__dict__")[name]
```
Comment (anmyachev) [inline modin/experimental/cloud/meta_magic.py:69]: Why you perform lookup inside `dict`?

File: modin/experimental/cloud/rpyc_proxy.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            return result
+
+        try:
+            # first of all, check if remote object has a known "wrapping" class
+            # example: _DataFrame has DataFrame dual-nature wrapper
+            local_cls = _KNOWN_DUALS[local_cls]
+        except KeyError:
+            pass
+        try:
+            wrapping_cls = object.__getattribute__(local_cls, "__dict__")[
```
Comment (anmyachev) [inline modin/experimental/cloud/rpyc_proxy.py:42]: Why `__getattribute__` performs on `__dict__`?

File: modin/pandas/dataframe.py (hunk 7 above)
Comment (anmyachev) [reply modin/pandas/dataframe.py:47]: Have you tried such a concept? `DataFrame.__new__` is a factory method and, depending on the context, returns a remote or local dataframe. To successfully execute methods of a remote dataframe, it is necessary to transfer local objects to the remote context using `deliver` function, but we also have the ability to get objects from local to remote using `obtain` function. To use it, you most likely we need to define another context that will be in the cluster when remote computing is performed. To manage requests, namely to be able to cache, we need to modify the `netref` class, maybe try to inherit from it, or try to use a special API inside `RPyC` for this(if exist).

File: modin/experimental/cloud/rpyc_proxy.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+            return DeliveringILocIndexer(self)
+
+    return DeliveringMixin
+
+
+def make_dataframe_wrapper():
+    from modin.pandas.dataframe import _DataFrame
+
+    DeliveringDataFrame = _deliveringWrapper(
+        _DataFrame, ["groupby", "agg", "aggregate"], _prepare_loc_mixin(), "DataFrame"
```
Comment (anmyachev) [inline modin/experimental/cloud/rpyc_proxy.py:193]: For taxi:q3 
```suggestion
        _DataFrame, ["groupby", "agg", "aggregate", "__getitem__"], _prepare_loc_mixin(), "DataFrame"
```

File: modin/data_management/factories.py (hunk 8 above)
Comment (devin-petersohn) [reply modin/data_management/factories.py:215]: I think we want all factory objects in the same place for simplicity. It allows us to go to one place for all supported combinations. Other `ExperimentalXFactory` objects live in this file as well.

File: modin/experimental/cloud/meta_magic.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        # "self" argument while it should hide it for our purposes.
+        # So we make a method bound to class type (it would normally be bound to instance)
+        # and pass that to .signature()
+        return inspect.signature(types.MethodType(self.__init__, self))
+
+    def __instancecheck__(self, instance):
+        # see if self is a type that has __real_cls__ defined,
+        # as if it's a dual-nature class we need to use its internal class to compare
+        try:
+            my_cls = self.__dict__["__real_cls__"]
```
Comment (devin-petersohn) [inline modin/experimental/cloud/meta_magic.py:42]: You can avoid this `try` `except` with the default.
```suggestion
            my_cls = self.__dict__.get("__real_cls__", self)
```

File: modin/pandas/test/test_api.py [hunk 14]
```diff
@@ -51,6 +51,7 @@ def test_top_level_api_equality():
         "base",
         "utils",
         "dataframe",
+        "groupby",
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/test/test_api.py:54]: Because there is no easy way to make a submodule _disappear_ from result of calling a `dir(module)`, and we need `modin.pandas.groupby` to be imported in the very beginning to RPyC to correctly make a netref class.

File: modin/pandas/io.py [hunk 15]
```diff
@@ -106,8 +106,8 @@ def parser_func(
         memory_map=False,
         float_precision=None,
     ):
-        _, _, _, kwargs = inspect.getargvalues(inspect.currentframe())
-        if kwargs.get("sep", sep) is False:
+        kwargs = dict(locals())
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/io.py:109]: I'm just going to revert this for now, this cleanup could be done later.

File: modin/pandas/dataframe.py [hunk 16]
```diff
@@ -457,8 +458,6 @@ def groupby(
                 elif mismatch:
                     raise KeyError(next(x for x in by if x not in self))
 
-        from .groupby import DataFrameGroupBy
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/dataframe.py:460]: Explained above. We need this to happen as early as possible for making a proper netref.

File: modin/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/__init__.py:100]: This is needed because order of callbacks starts to matter.

File: examples/cluster/mortgage-runner.py (hunk 2 above)
Comment (vnlitvinov) [reply examples/cluster/mortgage-runner.py:17]: Yes, indeed. I should fix the comment.

File: examples/cluster/mortgage-runner.py (hunk 3 above)
Comment (vnlitvinov) [reply examples/cluster/mortgage-runner.py:27]: agreed

File: examples/cluster/mortgage-runner.py (hunk 4 above)
Comment (vnlitvinov) [reply examples/cluster/mortgage-runner.py:29]: agreed

File: modin/__init__.py (hunk 5 above)
Comment (vnlitvinov) [reply modin/__init__.py:88]: As commented above, we need order, because we need to make sure a callback from `modin.pandas` happens first as it's the one that initializes remote Ray.

File: modin/experimental/cloud/meta_magic.py (hunk 9 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/meta_magic.py:18]: renamed

File: modin/experimental/cloud/meta_magic.py (hunk 10 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/meta_magic.py:69]: comment added

File: modin/experimental/cloud/rpyc_proxy.py (hunk 11 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/rpyc_proxy.py:42]: comment added

File: modin/experimental/cloud/rpyc_proxy.py (hunk 12 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/rpyc_proxy.py:193]: added

File: modin/experimental/cloud/meta_magic.py (hunk 13 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/meta_magic.py:42]: added

**Comment by vnlitvinov** [comment]
> Are you planning to split this?

In a sense. I've finished moving to a single class (i.e. something that doesn't require renaming foundation classes to `_Foo`), and I'm now looking into how to split the reviews properly into atomic pieces.

I think I'll use this PR eventually anyway, unless you have objections.

**Comment by devin-petersohn** [comment]
@vnlitvinov I have no objections, just checking 😄

**Comment by vnlitvinov** [comment]
Note to reviewers: I've started the process of chunking out atomic changes, so I'm converting this PR to draft _for now_.
It will come when https://github.com/modin-project/modin/pull/1805 and https://github.com/modin-project/modin/pull/1808 are merged, and I'll have to rebase it off `master` then and incorporate relevant changes.

Stay tuned 😃

**Comment by vnlitvinov** [comment]
Superseded by https://github.com/modin-project/modin/pull/1818
